source/unified-test-format/unified-test-format.rst

```diff
@@ -1342,6 +1365,65 @@ MUST assert that the field is not set, or, if the driver uses a nonpositive Int3
 value to indicate the field being unset, MUST assert that ``serverConnectionId``
 is a nonpositive Int32.
 
+expectedLogMessagesForClient
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A list of log messages that are expected to be observed (in that order) for a 
+client while executing `operations <test_operations_>`_.
+
+The structure of each object is as follows:
+
+- ``client``: Required string. Client entity on which the events are expected
+  to be observed. See `commonOptions_client`_.
+
+- ``messages``: Required array of `expectedLogMessage`_ objects. List of 
+  messages, which are expected to be observed (in this order) on the corresponding
+  client while executing `operations`_. If the array is empty, the test runner
+  MUST assert that no messages were observed on the client.
+
+expectedLogMessage
+~~~~~~~~~~~~~~~~~~
+
+A log message which is expected to be observed while executing the test's
+operations.
+
+The structure of each object is as follows:
+
+- ``level``: Required string. This MUST be one of the level names listed in
+   `log severity levels <logging/logging.rst#log-severity-levels>`__. This
+   specifies the expected level for the log message and corresponds to the
+   level used for the message in the specification that defines it. Note that
+   since not all drivers will necessarily support all log levels, some driver
+   may need to map the specified level to the corresponding driver-supported
+   level. Test runners MUST assert that the actual level matches this value.
+
+- ``component``: Required string. This MUST be one of the component names listed
+   in `components <../logging/logging.rst#components>`__. This specifies the
+   expected component for the log message. Note that since naming variations
+   are permitted for components, some drivers may need to map this to a
+   corresponding language-specific component name. Test runners MUST assert
+   that the actual component matches this value.
+
+- ``hasFailure``: Optional boolean. When this field is set to ``true``, 
```

Comment:
Since this has multiple conditions wrapped into a single boolean value, we should have tests for each cell in the truth table.  These are currently missing:

- assert `hasFailure: false for a non-exceptional log message (e.g. command-succeeded) 
- assert `hasFailure: true for an exceptional log message (e.g. command-failed with un-redacted failure) 

Also consider simplifying the boolean by making this one just about failure and a second one that is just about redaction.

Reply:
@jyemin and I discussed this offline and ended up having a broader conversation about error redaction. 

We reached the following conclusions:

* We should only redact command errors, not all errors. E.g. an IO exception in response to a sensitive command should not be redacted.
    * We can test this by using the failCommand failpoint with `closeConnection: true` to generate a non-command error.
* When we redact command errors, we should leave intact information that is definitely not going to be sensitive and could be helpful for debugging. We think [error code, error name, error labels] would be a reasonable allowlist. The error message and any other arbitrary fields should be redacted.
    * If a driver attaches the raw server response to command errors (as DRIVERS-2385 will mandate), this should be similarly redacted with only [error code, error name, error labels] present in the document.

An open question is what the right way to test the above redaction proposal is. We need to be able to test the following things when asserting a command error is redacted:
1. That an error code, error code name, and error labels are present/non-redacted.
2. That no other properties are present/unredacted.
3. That the attached server response only contains an error code, error code name, and error labels.

One potential path here is to adapt [expectedError](https://github.com/mongodb/specifications/blob/master/source/unified-test-format/unified-test-format.rst#expectederror) to support this, and add an optional `expectedError` to `expectedLogMessage`, which would replace `hasFailure` as currently defined in this PR.

Currently, `expectedError` would already allow us to make assertion 1 above. 

To support assertion 2, we could potentially add a new property to `expectedError` e.g. `isRedacted`,and when present the test runner would assert that all properties besides the allowed ones are null/empty. (I do worry that such an assertion might be rather brittle since it could be easy to add new properties to the command error type and forget to update this logic, but I don't really know a good way around that.)

To support assertion 3: `isRedacted` could also require asserting on the server response and checking it only contains certain keys. 
Alternatively, @jmikola and I discussed that DRIVERS-2385 would likely add an expected server response to `expectedError`. If that expected response is _not_ matched as a root document, the expected value could simply contain only keys from the allowlist to facilitate this. But, it does seem that in _most_ cases we would probably want treat this expected server response as a root document, similar to how the reply in a CommandSucceededEvent is always matched as a root document. That said, it always being root made DRIVERS-1653 and DRIVERS-1816 difficult and we ended up writing "best effort" tests that just check a subset of fields we'd expect to be present are not. I suppose the addition of `$$matchAsRoot` to the test format via this PR makes it possible for any non-root document to be matched as root when needed, so if we _did_ make it non-root by default there would be a way to opt into root matching when needed.

To summarize, the two options I'm proposing here are:
1. Add `isRedacted` to `expectedError`, and have `isRedacted` make assertions 2 and 3 above. 
2. Add `isRedacted` to `expectedError`, and have `isRedacted` assert that the error object has a code, codeName, and errorLabels, but no other properties. Have the expected server response being added in DRIVERS-2385 be matched as non-root by default, and use that for assertion 3. 

If you can think of other alternatives that seem better than either of these, let me know.

@jyemin let me know if I missed anything; @jmikola tagging you since this relates to DRIVERS-2385, and also for general input on the testing points.

Aside: the command monitoring equivalent ticket around testing redaction of failures is DRIVERS-2394. Ideally whatever we come up with here to test log message failure redaction would be mostly reusable there.

Reply:
Recapping the three assertions you presented above:

> 1. That an error code, error code name, and error labels are present/non-redacted.

Noted that this is already covered by existing syntax, as you said.

> 2. That no other properties are present/unredacted.

I agree that this does sound brittle. If you were to stick with a proprietary `hasFailure` on `expectedLogMessage`, does that avoid this concern at all?

> 3. That the attached server response only contains an error code, error code name, and error labels.

To my last question, I wonder if `expectedLogMessage.hasFailure` saying something to the effect of "the log message indicates a failure and MUST also assert that all fields other than error code, name, and labels are redacted."

Although even with that, asserting whether redaction have been made seems independent of whether the log indicates an error. I understand if there's really no other way to address that in test format syntax, though, without some additional field like `isFailureRedacted` that would always accompany `hasFailure` and be `true` (both verbose and redundant).

> Add isRedacted to expectedError, and have isRedacted make assertions 2 and 3 above.

Reading this, I'm not sure that allowing `expectedError` to be used within `expectedLogMessage` makes sense. Do you actually need to assert the logged error's specifics (e.g. code, name, labels) for this specification? Is that a nice-to-have? Or does the logging spec only care about whether there is an error and it's sufficiently redacted?

I'm wondering how other fields like `expectedError.expectResult` would work. I imagine the functionality isn't needed, but it'd be permitted by the schema and could complicate implementations. Likewise if other fields are later added to `expectedError` that don't make sense in the context of logging spec tests.

To that end, a basic `hasFailure` field with a loaded definition (and assertion logic) seems preferable, since it'd be fully contained to `expectedLogMessage` and less likely to conflict with anything else in a test runner.

----

> 2. Add isRedacted to expectedError, and have isRedacted assert that the error object has a code, codeName, and errorLabels, but no other properties. Have the expected server response being added in [DRIVERS-2385](https://jira.mongodb.org/browse/DRIVERS-2385) be matched as non-root by default, and use that for assertion 3.

My expectation (no pun intended) is that the server-side error document asserted for [DRIVERS-2385](https://jira.mongodb.org/browse/DRIVERS-2385) would be matched as a root document no different than the existing command/response assertions for APM events. I'll note that CommandFailedEvent doesn't have any assertion syntax for the `failure` property, but that's mainly because the Command Monitoring spec is very permissive about its type (i.e. string, exception, or document).

That agrees with your assumption from earlier in the comment:

> But, it does seem that in most cases we would probably want treat this expected server response as a root document, similar to how the reply in a CommandSucceededEvent is always matched as a root document.

----

> Aside: the command monitoring equivalent ticket around testing redaction of failures is [DRIVERS-2394](https://jira.mongodb.org/browse/DRIVERS-2394). Ideally whatever we come up with here to test log message failure redaction would be mostly reusable there.

I haven't given that ticket much thought before, but I imagine _that_ would be a case where a field like `isRedacted` would be necessary, since we don't have a reliable way to assert the possible types of CommandFailedEvent's attached error. In that case, `isRedacted` could be flexible enough to say "assert that whatever error type a driver uses to represented an error is redacted" and perhaps reference some common redaction rules already written for `expectedLogMessage.hasFailure`.

Reply:
So I went and tried to implement the "partial" (i.e. only redact certain properties) redaction discussed above in the Rust driver today. One thing I ran into is that, in our tracing events (essentially structured log messages) we can only attach primitive types to messages - strings plus a handful of numeric types. This means that errors will have to be represented as strings. Note that I already discuss that errors can be represented flexibly in log messages in the spec - see section "Representing Errors in Log Messages" - which is consistent with command monitoring.

In addition to the points you make above, I think this is another argument against my previous proposal to use `expectedError`, since we would not actually have an error object to assert against anyway. I also imagine that we are not the only drivers that might find ourselves in this situation. 

> Do you actually need to assert the logged error's specifics (e.g. code, name, labels) for this specification? Is that a nice-to-have? Or does the logging spec only care about whether there is an error and it's sufficiently redacted?

I would consider it more a nice-to-have than a necessity. We could assert on the error more thoroughly by expecting an error for the corresponding operation. Asserting presence and redaction/lack thereof seems likely good enough.

> Although even with that, asserting whether redaction have been made seems independent of whether the log indicates an error. I understand if there's really no other way to address that in test format syntax, though, without some additional field like isFailureRedacted that would always accompany hasFailure and be true (both verbose and redundant).

I think there is actually a use case for isFailureRedacted=false, which I mentioned above: non-command errors should _not_ be redacted. Maybe we should circle back around to Jeff's proposal at the beginning of this thread:

> Consider simplifying the boolean by making this one just about failure and a second one that is just about redaction.

We could use two booleans:
* `hasFailure`: this would probably only ever be true; when present we assert there is a failure attached to the message
* `isFailureRedacted`: this should only ever be specified when `hasFailure=true`. if false, assert failure is _not_ redacted (we would use this to test e.g. network errors are not redacted); if true assert failure _is_ redacted. 

This does still leave us with a couple problems: 
* brittle assertions around whether an error actually got redacted - a driver could forget to update the runner to assert that some new field is redacted once it is added. That said, I'm not sure drivers adding new fields to their `CommandError` types is a common occurrence at all (probably the last time was when error labels came to exist?) so this might not be worth worrying about.
* difficulty of asserting redaction for languages that represent errors as strings. In Rust for example an error string representation might look something like `CommandError { code: 1, code_name: "CodeNameHere", message: "Message here", labels: [] }` and so to assert redaction we'd have to do some kind of string parsing that verifies particular parts of that string are not there / replaced by some placeholder. That said, that doesn't seem like a big issue. We would only be using this on a single error type for now (command errors) and writing a regex to match only a command error and parse out its constituent parts would be straightforward. 

So I think I'm inclined to go something more like that route now. What do you think? @jyemin  / @BorisDog, how will the errors you include in log messages be represented (for Java, what is the intermediate representation before stringifying?) and would it be feasible to assert redaction/lack of redaction on that representation?

Reply:
I missed @jyemin's original suggestion when responding to your previous comment, but having two separate fields seems entirely reasonable. You should also be able to enforce that `isFailureRedacted` is only specified when `hasFailure: true` is specified -- I believe that would be similar to how we assert that an `operation` doesn't mix `expectError` with `expectResult`/`saveResultAsEntity`.

> brittle assertions around whether an error actually got redacted

I may be missing something else, but is there any proposal that avoids this concern? It seems like it'd always be an issue and there's no foolproof way to prevent a driver from making a CVE-worthy mistake.

> difficulty of asserting redaction for languages that represent errors as strings

This may be another misunderstanding on my part, but I also don't see how that can be avoided if a driver is limited to using a string type and the tests ultimately need to assert that redaction has taken place.

Reply:
changes have been pushed to switch to the two-boolean approach (see my top-level reply below for some related details)

Reply:
LGTM, but I won't resolve this thread myself since there are so many commenters.